src/DotNetWorker.Grpc/Handlers/InvocationHandler.cs

```diff
@@ -0,0 +1,154 @@
+using System.Threading.Tasks;
+using Azure.Core.Serialization;
+using Microsoft.Azure.Functions.Worker.Context.Features;
+using Microsoft.Azure.Functions.Worker.Grpc;
+using Microsoft.Azure.Functions.Worker.Grpc.Features;
+using Microsoft.Azure.Functions.Worker.Grpc.Messages;
+using Microsoft.Azure.Functions.Worker.Logging;
+using Microsoft.Azure.Functions.Worker.OutputBindings;
+using Microsoft.Azure.Functions.Worker.Rpc;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Functions.Worker.Handlers
+{
+    internal class InvocationHandler : IInvocationHandler
+    {
+        private readonly IFunctionsApplication _application;
+        private readonly IInvocationFeaturesFactory _invocationFeaturesFactory;
+        private readonly IOutputBindingsInfoProvider _outputBindingsInfoProvider;
+        private readonly IInputConversionFeatureProvider _inputConversionFeatureProvider;
+        private readonly ObjectSerializer _serializer;
+        private readonly ILogger _logger;
+
+        private ConcurrentDictionary<string, CancellationTokenSource> _inflightInvocations => new();
+
+        public InvocationHandler(
+            IFunctionsApplication application,
+            IInvocationFeaturesFactory invocationFeaturesFactory,
+            ObjectSerializer serializer,
+            IOutputBindingsInfoProvider outputBindingsInfoProvider,
+            IInputConversionFeatureProvider inputConversionFeatureProvider,
+            ILoggerProvider loggerProvider)
+        {
+            _application = application ?? throw new ArgumentNullException(nameof(application));
+            _invocationFeaturesFactory = invocationFeaturesFactory ?? throw new ArgumentNullException(nameof(invocationFeaturesFactory));
+            _serializer = serializer ?? throw new ArgumentNullException(nameof(serializer));
+            _outputBindingsInfoProvider = outputBindingsInfoProvider ?? throw new ArgumentNullException(nameof(outputBindingsInfoProvider));
+            _inputConversionFeatureProvider = inputConversionFeatureProvider ?? throw new ArgumentNullException(nameof(inputConversionFeatureProvider));
+            _logger = loggerProvider.CreateLogger(nameof(InvocationHandler));
+        }
+
+        public async Task<InvocationResponse> InvokeAsync(InvocationRequest request)
+        {
+            FunctionContext? context = null;
+            InvocationResponse response = new()
+            {
+                InvocationId = request.InvocationId,
+                Result       = new StatusResult()
+            };
+
+            try
+            {
+                var invocation = new GrpcFunctionInvocation(request);
+                var cancellationTokenSource = new CancellationTokenSource();
+
+                if (!_inflightInvocations.TryAdd(request.InvocationId, cancellationTokenSource))
+                {
+                    // Dispose the CancellationTokenSource if we cannot keep track of it
+                    cancellationTokenSource?.Dispose();
```

Comment:
The code as is will continue invocation of this function if we cannot track this token. This will lead to a couple of issues:

1. The `CancellationTokenSource` is now disposed, so accessing the `.Token` below will throw.
2. Even if we solve the object disposal issue, what will the expected behavior for a gRPC cancel message now? That CTS will not be correctly tracked in the dictionary. So we will potentially cancel the _wrong_ CTS and not cancel the correct one at all.

Do you see any valid reason why we should continue invocation if we fail to track the CTS? If so, how can we reconcile point 2 above? If we can't come up with a good solution to the above, it might be better to just treat this as an error case and fail the invocation. The alternative may cause some very obscure bugs.

Reply:
Yeah this is an interesting one, the part where I'm conflicted is if a customer is not binding to a cancellation token but we fail because we can't save the CTS (that we may never use?); is it valid for us to fail an entire invocation based on that?

Reply:
I think we should _always_ have a CTS for any invocation. Whether the end function binds that parameter or not, it should always be available to middleware to respond to. I believe this is what @fabiocav was mentioning.